frontend/lib/src/components/shared/Profiler/Profiler.tsx

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {
+  FC,
+  ProfilerOnRenderCallback,
+  PropsWithChildren,
+  Profiler as ReactProfiler,
+} from "react"
+
+import { CircularBuffer } from "./CircularBuffer"
+
+export type ProfilerProps = PropsWithChildren<{
+  /**
+   * The unique ID of a Profiler.
+   * Statically typed so we can enforce which profiles are being collected.
+   */
+  id: "Main" | "Sidebar" | "Bottom" | "Event"
+}>
+
+/**
+ * Callback function for React Profiler that collects performance data and
+ * writes it to the global `window.__streamlit_profiles__` object so that it can
+ * be collected in our performance tests.
+ */
+const handleRender: ProfilerOnRenderCallback = (
+  id,
+  phase,
+  actualDuration,
+  baseDuration,
+  startTime,
+  commitTime
+) => {
+  window.__streamlit_profiles__ = window.__streamlit_profiles__ || {}
+
+  window.__streamlit_profiles__[id] =
+    window.__streamlit_profiles__[id] ||
+    // Use a CircularBuffer to limit the number of profiles stored in memory to
+    // prevent any potential memory leaks.
+    new CircularBuffer<(typeof window.__streamlit_profiles__)[string]>(1000)
```

Comment:
I assume `1000` was arbitrarily chosen (which is fine!) but I think a comment would be great to state so. Otherwise someone might look at this in a year from now and is hesitant to touch the number because they do not know why it was set to it in the first place.